tests/test_attributes.py

```diff
@@ -70,7 +70,20 @@ def test_attributes(browser, netid, utils, sp_url, sp_domain, secrets, test_env)
         content_clean = content.get_attribute('innerHTML')
         attribute_data = [item.strip() for item in content_clean.split("<br>")]
 
+        # eduPersonScopedAffiliation is a unique case
+        wanted_key = 'eduPersonScopedAffiliation'
+        result = list(filter(lambda x: x.startswith(wanted_key), attribute_data))
+        found_values = result[0].split("= ", 1)[1].split(";")
+
         for key, value in attributes_to_test.items():
             # matches on exact string only. mail won't count as a match for uwEduEmail and uwEduEmail won't match mail
-            assert (f'{key} = {value}' in attribute_data)
+            if key != 'eduPersonScopedAffiliation':
+                assert (f'{key} = {value}' in attribute_data)
+            # eduPersonScopedAffiliation values may not be in the same order every time.
+            elif key == 'eduPersonScopedAffiliation':
+                assert len(found_values) == 4
+                assert 'employee@washington.edu' in found_values
+                assert 'student@washington.edu' in found_values
+                assert 'staff@washington.edu' in found_values
+                assert 'member@washington.edu' in found_values
```

Comment:
A few annotated suggestions here to reduce repetition and complexity:

```
result = next(filter(lambda x: x.startswith(wanted_key)))  # wrap with `next`
found_values = set(result.split("= ", limit=1)[1].split(";")   # wrap with `set`
for key, value in attributes_to_test.items():
    if key == wanted_key:
        assert f'{key} = {value}' in attribute_data
   else:
       assert found_values == {  # Compare equality of two sets
           'employee@washington.edu',
           'student@washington.edu',
           'staff@washington.edu',
           'member@washingtone.edu',
       }
```

**Why `next`?**

The current version creates an iterator and iterates through all items even though it only wants the first item in the list. Replacing `list` with `next` will fetch that first item for you and save the processing of other items that you don't need.

**Why `if/else`?**

What you have is already if/else, only it's an if/else construct that is open to someone making a mistake and requires someone to modify a a string in 3 places if it needs to change. Since you've already defined `wanted_key`, you may as well just use it.

**Why a `set`**?

You currently test `len == 4` followed by exactly 4 tests for the contents This tells me that what you have currently is a list of undetermined order. That's what a set is. So, better to use an existing data type exactly for its purpose rather than perform extra work for the wrong data type. Now, if we ever need to update these values, all we need to do is add the item, instead of adding a new assertion _and_ updating the expected count.

Reply:
`if key == wanted_key:`
I think you meant `if key != wanted_key:`

`result = next(filter(lambda x: x.startswith(wanted_key)))  # wrap with next`
cool, so the loop will stop when it finds the first occurrence?
I guess I still need the `attribute_data` param though

> Why if/else?

I put `elif key == 'eduPersonScopedAffiliation':` there anticipating more keys that aren't listed here will be added, just safeguarding to when we get there.

> Why a set? 

Cool.

Reply:
>  there anticipating more keys that aren't listed here will be added, just safeguarding to when we get there.

If that condition doesn't exist right now, there's no reason to safeguard for it, it only makes the code make less sense until that time.

Reply:
>  I think you meant if key != wanted_key:

No, I meant what I said.

"if x is not y, then do z, otherwise to w"
is the same as "if x is y, do w, otherwise do z"

but it avoids a negative expression, which just adds  more complexity.

Reply:
(again, these are recommendations, not mandates, just evangelizing)

Reply:
Ah, so the if/else, I just needed to swap the if/else bodies (since the special case is the wanted key), but still keep the conditions you suggested.
for key, value in attributes_to_test.items():
```
   if key == wanted_key:
       assert found_values == {  # Compare equality of two sets
           'employee@washington.edu',
           'student@washington.edu',
           'staff@washington.edu',
           'member@washingtone.edu',
       }
   else:  # No reason to run a second comparison; either the value is the wanted_key or it isn't.
        assert f'{key} = {value}' in attribute_data
```

I'll edit the code and get this re-pushed. I appreciate the input.

Reply:
no problem! Thanks for following up!